### DISCUSSION TITLE
Getting started with Falcon

### DISCUSSION THREAD

**Discussion by nepthar** [socketry/falcon#149] [Q&A] [answered]
Hi folks, I'm currently a RoR developer, but fairly new to the Ruby ecosystem overall. I'm trying to evaluate falcon for some internal services that have to be a bit more performant that Rails and I think that this fits the bill.

I've been through all of your documentation but I'm still struggling to get Falcon up and running on my dev environment (macOS 11.2.3). For instance, the getting started guide says run `falcon serve`, but I'm not quite sure how to interpret that (should I copy `bin/falcon` into my PATH? do I clone the falcon repo for each new project?).

Here's what I've tried:
1. Clone the repo
2. `rvm install ruby-3.0.0` & `rvm use ruby-3.0.0`
3. `cd falcon`, `bundle install` Looks like it installs some dependencies correctly?
4. `./bin/falcon serve` Looks like it gets somewhere, but then errors out with "Too many open files", so I suspect something is wrong/it's opening files in a tight loop.

I've also tried to just run some of the examples directly:
1. `cd examples/hello`
2. `../../bin/falcon serve`

That errors with the same message. I feel that I'm missing some basic understanding about how this works. Is there some additional documentation you can point me to or advice you can give on how to get up and running? 

I'm happy to submit a PR to expand the getting started documentation once I figure it out. Thanks!

**Comment by havenwood** [comment]
"Too many open files" `ulimit` limitation is mentioned in the documentation but it's under the [Performance Tuning](https://github.com/socketry/falcon/blob/v0.38.1/guides/performance-tuning/README.md#system-limitations) guide. For example, you can temporarily bump macOS's low maxfiles with:

```sh
sudo launchctl limit maxfiles 65536 unlimited
```

Check into how to persist `ulimit` settings on your system.

**Reply by ioquatix** [reply]
To be honest, I'm surprised you'd hit this in a development environment. Falcon normally only hits this issue if you are doing some kind of stress test and creating thousands of persistent connected clients.

**Reply by nepthar** [reply]
Yep, hitting ulimit in development for a simple project is likely an indication that I'm doing something very, very wrong, which I was :)

**Comment by ioquatix** [comment, marked as answer]
You should add falcon as a gem to your project and then use `bundle exec falcon serve`. You shouldn't be cloning the git repo to run it in development or production.

**Reply by nepthar** [reply]
Okay, I understand now. I really was just missing ruby ecosystem basics. I didn't know I had to create a new project with a Gemfile and config.ru file, but putting that all together things seem to work, thanks

**Reply by ioquatix** [reply]
The examples should also just run directly from the repo, provided you do bundle install first. This sets up the environment. Bundle will typically pull the falcon code from the local repo if you are within the examples directory, but not always.
